{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "## 1. Genome design (Genotype construction)\n",
    "- Node genes: neuron in the neural network\n",
    "- Connection genes: connections between nodes, each with innovation numbers\n",
    "- Genomes: a complete neural network (phenotype) that stores all child connection and node genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import pyglet\n",
    "from pyglet import shapes\n",
    "import time\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "represents a capability of the system \n",
    "ex: firing capability, wind speed\n",
    "\"\"\"\n",
    "class NodeGene:\n",
    "    def __init__(self, id, layer, activation, bias):\n",
    "        self.id = id\n",
    "        self.layer = layer # layer node belongs to (input, hidden, output)\n",
    "        self.activation = activation\n",
    "        self.bias = bias\n",
    "\n",
    "\"\"\"\n",
    "represents a rule that utilizes the capability\n",
    "ex: wind speed (input node A) increases (connection) firing capability (output node B)\n",
    "\"\"\"\n",
    "class NodeConnection:\n",
    "    def __init__(self, id, output_node_id, input_node_id, weight, enabled, innovation):\n",
    "        self.id = id\n",
    "        self.output_node_id = output_node_id\n",
    "        self.input_node_id = input_node_id\n",
    "        self.weight = weight # learned\n",
    "        self.enabled = enabled # disabled gene?\n",
    "        self.innovation = innovation # id tracks when the connection first showed up in evolutionary history\n",
    "\n",
    "\"\"\"\n",
    "represents the entire rule book for a team\n",
    "ex: doctrine manual for red team vs blue team or a single vessel\n",
    "\"\"\"\n",
    "class Genome:\n",
    "    def __init__(self, nodes, connections):\n",
    "        self.nodes = {node.id: node for node in nodes} # each node keyed by id\n",
    "        self.connections = {conn.innovation: conn for conn in connections} # each connection is keyed by innovation\n",
    "        self.fitness = 0\n",
    "    \n",
    "    def __repr__(self):\n",
    "        return f\"Genome(nodes={len(self.nodes)}, connections={len(self.connections)}, fitness={self.fitness})\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3",
   "metadata": {},
   "source": [
    "## 2. Crossover algorithm (heritage)\n",
    "Perform crossover between two parent genomes to derive a new child genome that inherits structures from the parent\n",
    "- matching genes: randomize which parent's genes are expressed. topological connection is the same but strength (weight) differs\n",
    "- disjoint gene: \n",
    "- excess genes: -"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\"\"\"\n",
    "finds the child connection set\n",
    "input: parent_a and parent_b genomes\n",
    "output: child genome\n",
    "\"\"\"\n",
    "def crossover(parent_a, parent_b, population):\n",
    "    # ensure parents are different\n",
    "    if parent_a == parent_b:  # Or check if weights are identical\n",
    "        parent_a = random.choice([g for g in population if g != parent_b])\n",
    "    \n",
    "    parent_fittest = find_fittest_parent(parent_a, parent_b)\n",
    "\n",
    "    # get set of innovation numbers of parents\n",
    "    inn_num_parent_a = set(parent_a.connections.keys())\n",
    "    inn_num_parent_b = set(parent_b.connections.keys())\n",
    "\n",
    "    matching = inn_num_parent_a & inn_num_parent_b\n",
    "    only_in_a = inn_num_parent_a - inn_num_parent_b\n",
    "    only_in_b = inn_num_parent_b - inn_num_parent_a\n",
    "\n",
    "    # iterate over all innovation numbers\n",
    "    # collect the node connections in child\n",
    "    child_node_connections = []\n",
    "    \n",
    "    for innovation in matching | only_in_a | only_in_b:\n",
    "        # 1. matching nums: random selection of gene\n",
    "        if (innovation in matching):\n",
    "            random_connection = random.choice([parent_a.connections[innovation], parent_b.connections[innovation]])\n",
    "\n",
    "            # create new conn obj for the child from parents\n",
    "            child_conn = NodeConnection(\n",
    "                id=random_connection.id,\n",
    "                output_node_id=random_connection.output_node_id,\n",
    "                input_node_id=random_connection.input_node_id,\n",
    "                weight=random_connection.weight,  # inherit from selected parent\n",
    "                enabled=random_connection.enabled,\n",
    "                innovation=random_connection.innovation\n",
    "            )\n",
    "            \n",
    "            child_node_connections.append(child_conn)\n",
    "        # 2. include if disjoint/excess in fitter parent\n",
    "        elif (innovation in parent_fittest.connections):\n",
    "            parent_conn = parent_fittest.connections[innovation]\n",
    "\n",
    "            # create new conn obj for the child from parents\n",
    "            child_conn = NodeConnection(\n",
    "                id=parent_conn.id,\n",
    "                output_node_id=parent_conn.output_node_id,\n",
    "                input_node_id=parent_conn.input_node_id,\n",
    "                weight=parent_conn.weight,  # inherit from selected parent\n",
    "                enabled=parent_conn.enabled,\n",
    "                innovation=parent_conn.innovation\n",
    "            )\n",
    "            \n",
    "            child_node_connections.append(child_conn)\n",
    "        # 3. dont include if disjoint/excess in less fit parent\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "    # collect child nodes expressed\n",
    "    child_node_ids = set()\n",
    "    for conn in child_node_connections:\n",
    "        child_node_ids.add(conn.input_node_id)\n",
    "        child_node_ids.add(conn.output_node_id)\n",
    "    \n",
    "    # create a new child node for each child\n",
    "    child_nodes = [\n",
    "        NodeGene(\n",
    "        id=node_id,\n",
    "        layer=parent_fittest.nodes[node_id].layer,\n",
    "        activation=parent_fittest.nodes[node_id].activation,\n",
    "        bias=parent_fittest.nodes[node_id].bias\n",
    "        )\n",
    "        for node_id in child_node_ids\n",
    "    ]\n",
    "\n",
    "    # build child genome\n",
    "    child = Genome(\n",
    "        nodes = child_nodes,\n",
    "        connections = child_node_connections\n",
    "    )\n",
    "\n",
    "    return child\n",
    "\n",
    "\"\"\"\n",
    "finds the fitter parent based on fitness score\n",
    "input: parent_a and parent_b genomes\n",
    "output: fitter parent genome\n",
    "\"\"\"\n",
    "def find_fittest_parent(a, b):\n",
    "    return a if a.fitness >= b.fitness else b\n",
    "\n",
    "\"\"\"\n",
    "generates a graph visualization of genome\n",
    "\"\"\"\n",
    "def visualize_genome(genome, ax=None):\n",
    "    G = nx.DiGraph()\n",
    "\n",
    "    # create edge for each enabled connection\n",
    "    for conn in genome.connections.values():\n",
    "        if conn.enabled:\n",
    "            G.add_edge(conn.input_node_id, conn.output_node_id)\n",
    "    \n",
    "    # draw force directed graph\n",
    "    if ax is None:\n",
    "        nx.draw(G, with_labels=True, arrows=True)\n",
    "        plt.show()\n",
    "    else:\n",
    "        nx.draw(G, ax=ax, with_labels=True, arrows=True)\n",
    "\n",
    "\"\"\"\n",
    "print genome state\n",
    "\"\"\"\n",
    "def print_genome_state(genome, label=\"Genome\"):\n",
    "    print(f\"\\n=== {label} ===\")\n",
    "    print(f\"Nodes: {list(genome.nodes.keys())}\")\n",
    "    print(\"Connections:\")\n",
    "    for inn, conn in genome.connections.items():\n",
    "        status = \"enabled\" if conn.enabled else \"DISABLED\"\n",
    "        print(f\"  [{inn}] {conn.input_node_id} → {conn.output_node_id} | weight: {conn.weight:.4f} | {status}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5",
   "metadata": {},
   "source": [
    "## 3. Innovation counter\n",
    "tracks the evolutionary history of all new genetic mutations that are created over time for a set population\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "class InnovationCounter:\n",
    "    def __init__(self, global_innovations, node_connection_innovations):\n",
    "        self.global_innovations = global_innovations\n",
    "        self.node_connection_innovations = node_connection_innovations or {} # dict (key(in_node_id, out_node_id) -> innovation number)\n",
    "    \n",
    "    \"\"\"\n",
    "    get innovation number per connection. create new one if it doesn't exist\n",
    "    tracks new unique strategies developed\n",
    "    \"\"\"\n",
    "    def get_inn_for_connection(self, in_node_id, out_node_id):\n",
    "        # key by in/out node\n",
    "        key = (in_node_id, out_node_id)\n",
    "\n",
    "        # assign or create new innovation\n",
    "        if (key in self.node_connection_innovations):\n",
    "            return self.node_connection_innovations[key]\n",
    "        else:\n",
    "            # increment global counter\n",
    "            self.global_innovations += 1\n",
    "\n",
    "            # create new innovation number\n",
    "            self.node_connection_innovations[key] = self.global_innovations\n",
    "            return self.global_innovations\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7",
   "metadata": {},
   "source": [
    "## 4. Mutations\n",
    "apply noise to children after crossover for natural evolution\n",
    "- mutation_rate - percentage of population mutated\n",
    "- perturbation_strength - weight is slightly changes\n",
    "- reset_rate - weight fully randomized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "update execution intensity of particular strategies\n",
    "\"\"\"\n",
    "def mutate_weights(genome, mutation_rate, perturbation_strength, reset_rate):\n",
    "    for conn in genome.connections.values():\n",
    "        # run mutation\n",
    "        if random.random() < mutation_rate:\n",
    "            # determine type of mutation\n",
    "            if random.random() < reset_rate:\n",
    "                conn.weight = random.uniform(0, 1)\n",
    "            else:\n",
    "                # apply slight gaussian perturbation\n",
    "                conn.weight += random.gauss(0, perturbation_strength)\n",
    "\n",
    "# \"\"\"\n",
    "# create new units or variables (ex: sensor unit, wind speed)\n",
    "# \"\"\"\n",
    "# function mutate_nodes(genome, innov: InnovationCounter):\n",
    "\n",
    "\"\"\"\n",
    "create new connection between two nodes\n",
    "\"\"\"\n",
    "def mutate_connections(genome, mutation_rate, innov: InnovationCounter):\n",
    "    # get all possible node pairs\n",
    "    # possible pairs do not create cyclical connections\n",
    "    all_node_ids = list(genome.nodes.keys())\n",
    "    possible_node_pairs = list([(a, b) for a in all_node_ids for b in all_node_ids if a != b ])\n",
    "    \n",
    "    # filter out existing node pairs\n",
    "    existing_connections = set((conn.input_node_id, conn.output_node_id) for conn in genome.connections.values())\n",
    "\n",
    "    # valid pairs are not already existing\n",
    "    valid_pairs = [\n",
    "        (a, b) \n",
    "        for (a, b) in possible_node_pairs \n",
    "        if (a,b) not in existing_connections\n",
    "        if genome.nodes[a].layer < genome.nodes[b].layer # lower -> higher layer connections only\n",
    "    ]\n",
    "\n",
    "    if not valid_pairs:\n",
    "        return\n",
    "\n",
    "    # select 2 random pairs\n",
    "    if random.random() < mutation_rate:\n",
    "        (in_node, out_node) = random.choice(valid_pairs)\n",
    "\n",
    "    # create new connection and add to genome\n",
    "    innovation_num = innov.get_inn_for_connection(in_node, out_node)\n",
    "    \n",
    "    new_conn = NodeConnection(\n",
    "        id=1, \n",
    "        input_node_id=in_node, \n",
    "        output_node_id=out_node, \n",
    "        weight=random.uniform(0,1), \n",
    "        enabled=True, \n",
    "        innovation=innovation_num\n",
    "    )\n",
    "\n",
    "    genome.connections[innovation_num] = new_conn"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9",
   "metadata": {},
   "source": [
    "#### 4a. test crossover x mutations algorithm\n",
    "expected: child should adopt all of parent_a's traits (fitter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# nodes\n",
    "node_1 = NodeGene(id=1, layer=0, activation='sigmoid', bias=0) # input\n",
    "node_2 = NodeGene(id=2, layer=0, activation='sigmoid', bias=0) # input\n",
    "\n",
    "node_3 = NodeGene(id=3, layer=1, activation='sigmoid', bias=0) # hidden\n",
    "node_4 = NodeGene(id=4, layer=2, activation='sigmoid', bias=0) # output\n",
    "\n",
    "# create connections for parent A\n",
    "conn_a1 = NodeConnection(id=1, output_node_id=3, input_node_id=1, \n",
    "                        weight=0.87, enabled=True, innovation=1)\n",
    "conn_a2 = NodeConnection(id=2, output_node_id=4, input_node_id=3, \n",
    "                        weight=0.5, enabled=True, innovation=2)\n",
    "\n",
    "# create connections for parent B  \n",
    "conn_b1 = NodeConnection(id=1, output_node_id=3, input_node_id=1, \n",
    "                        weight=0.5, enabled=True, innovation=1)\n",
    "conn_b3 = NodeConnection(id=3, output_node_id=4, input_node_id=1, \n",
    "                        weight=0.5, enabled=True, innovation=3)\n",
    "\n",
    "# build genomes\n",
    "parent_a = Genome(nodes=[node_1, node_2, node_3, node_4], \n",
    "                connections=[conn_a1, conn_a2])\n",
    "parent_a.fitness = 10\n",
    "\n",
    "parent_b = Genome(nodes=[node_1, node_2, node_3], \n",
    "                connections=[conn_b1, conn_b3])\n",
    "parent_b.fitness = 5\n",
    "\n",
    "population = [parent_a, parent_b]\n",
    "\n",
    "# init counter\n",
    "innov = InnovationCounter(\n",
    "    global_innovations = 2, # start after existing innovations\n",
    "    node_connection_innovations = {(1, 3): 1, (2, 3): 2}  # dict with existing connections\n",
    ")\n",
    "\n",
    "# print parents\n",
    "fig, axes = plt.subplots(1, 4, figsize=(15, 5))  # 1 row, 4 columns\n",
    "\n",
    "print_genome_state(parent_a, \"Parent A (fitness=10)\")\n",
    "visualize_genome(parent_a, ax=axes[0])\n",
    "axes[0].set_title(\"Parent A (fitness=10)\")\n",
    "\n",
    "print_genome_state(parent_b, \"Parent B (fitness=5)\")\n",
    "visualize_genome(parent_b, ax=axes[1])\n",
    "axes[1].set_title(\"Parent B (fitness=5)\")\n",
    "\n",
    "# create and print crossover\n",
    "child = crossover(parent_a, parent_b, population)\n",
    "print_genome_state(child, \"Before Mutations\")\n",
    "visualize_genome(child, ax=axes[2])\n",
    "axes[2].set_title(\"Child before mutation\")\n",
    "\n",
    "# mutate weights\n",
    "mutate_weights(child, mutation_rate=1.0, perturbation_strength=0.3, reset_rate=0.1)\n",
    "mutate_connections(child, mutation_rate=1.0, innov=innov)\n",
    "print_genome_state(child, \"After Mutations\")\n",
    "visualize_genome(child, ax=axes[3])\n",
    "axes[3].set_title(\"Child after mutation\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11",
   "metadata": {},
   "source": [
    "## 5. Evolutionary loop (Phenotype construction)\n",
    "apply crossover over generations of genomes until population fitness threshold is met\n",
    "\n",
    "Core loop:\n",
    "1. evaluation: forward pass and evaluate against fitness function\n",
    "2. speciation: nodes with similar topologies are chosen to reproduce\n",
    "3. selection: those with high fitness scores are selected to reproduce\n",
    "4. reproduction: crossover and mutation across selected pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "class EvolutionLoop:\n",
    "    \"\"\"\n",
    "    track the fitness and innovation numbers for the entire population\n",
    "    \"\"\"\n",
    "    def __init__(self, innov):\n",
    "        # self.fitness_fn = fitness_fn\n",
    "        self.innov = innov\n",
    "\n",
    "    \"\"\"\n",
    "    sort genome based on input->hidden->output structure\n",
    "    \"\"\"\n",
    "    def topological_sort(self, genome):\n",
    "        # compute the in-degree of each node\n",
    "        in_degree = {id: 0 for id in genome.nodes.keys()}\n",
    "        for conn in genome.connections.values():\n",
    "            if conn.enabled:\n",
    "                in_degree[conn.output_node_id] += 1\n",
    "        \n",
    "        # create a processing queue of nodes with no incoming edges (fifo)\n",
    "        queue = [nid for nid, deg in in_degree.items() if deg == 0]\n",
    "        topo_order = []\n",
    "\n",
    "        # process each node in order \n",
    "        while queue:\n",
    "            curr_node = queue.pop(0)\n",
    "            topo_order.append(curr_node)\n",
    "\n",
    "            # check next node\n",
    "            for conn in genome.connections.values():\n",
    "                if conn.enabled and conn.input_node_id == curr_node:\n",
    "                    out = conn.output_node_id # find the output node\n",
    "                    in_degree[out] -= 1 # decrement the output node\n",
    "                    if in_degree[out] == 0: # add to queue\n",
    "                        queue.append(out)\n",
    "        \n",
    "        # check for cycles\n",
    "        if len(topo_order) != len(genome.nodes):\n",
    "            raise ValueError(\"Cycle detected\")\n",
    "        \n",
    "        return topo_order\n",
    "    \n",
    "    \"\"\"\n",
    "    given an input state of the world, compute the resulting actions\n",
    "    - genome: node structure\n",
    "    - input_values: external node values fed into the network\n",
    "    \"\"\"\n",
    "    def forward_pass(self, genome, input_values):\n",
    "        # collect intermediate values of each node in the network\n",
    "        node_values = {}\n",
    "\n",
    "        # map input values to input nodes\n",
    "        for nid, val in input_values.items():\n",
    "            if nid in genome.nodes:  # only process nodes that exist in this genome\n",
    "                node_values[nid] = val\n",
    "        \n",
    "        # topological sort of nodes\n",
    "        ordered_nodes = self.topological_sort(genome)\n",
    "\n",
    "        # process each node in the network\n",
    "        for nid in ordered_nodes:\n",
    "            # skip all input values and nodes\n",
    "            if nid in input_values and genome.nodes[nid].layer == 0:\n",
    "                continue\n",
    "        \n",
    "            # incoming_sum = bias + Σ (source_value × weight)\n",
    "            incoming_sum = genome.nodes[nid].bias\n",
    "            for conn in genome.connections.values():\n",
    "                if conn.enabled and conn.output_node_id == nid:\n",
    "                    source_value = node_values[conn.input_node_id]\n",
    "                    incoming_sum += source_value * conn.weight\n",
    "            \n",
    "            # apply activation function to sum (sigmoid)\n",
    "            normalized_sum = max(-700, min(700, incoming_sum))\n",
    "            node_values[nid] = 1 / (1 + math.exp(-normalized_sum))\n",
    "\n",
    "        # return output nodes from the full pass\n",
    "        return [\n",
    "            node_values[nid] for nid in node_values\n",
    "            if nid in genome.nodes and genome.nodes[nid].layer == 2 \n",
    "        ]\n",
    "\n",
    "    \"\"\"\n",
    "    compute fitness based on the lanchester square equation\n",
    "    rewards greatest units number x firepower strength\n",
    "    - input: current env state\n",
    "    \"\"\"\n",
    "    # \\(dA/dt=-bB\\) and \\(dB/dt=-aA\\),\n",
    "    def calculate_fitness(self, obs):\n",
    "        blue_units = obs[1]\n",
    "        red_units = obs[2]\n",
    "\n",
    "        # blue preserved\n",
    "        units_survived = blue_units / 27\n",
    "\n",
    "        # red eliminated\n",
    "        enemy_eliminated = (33 - red_units) / 33\n",
    "\n",
    "        # victory condition (lanchester)\n",
    "        blue_combat_power = obs[4] * (blue_units ** 2)\n",
    "        red_combat_power = obs[5] * (red_units ** 2)\n",
    "        victory_bonus = 1.0 if blue_combat_power > red_combat_power else 0.0\n",
    "\n",
    "        # compute weighted combination\n",
    "        fitness = (\n",
    "            0.4 * units_survived +      # Preserve your forces\n",
    "            0.3 * enemy_eliminated +     # Eliminate enemy\n",
    "            0.2 * victory_bonus          # Victory condition\n",
    "        )\n",
    "\n",
    "        return fitness\n",
    "\n",
    "    \"\"\"\n",
    "    create initial genome state\n",
    "    \"\"\"\n",
    "    def create_initial_genome(self, num_inputs, num_outputs, innov):\n",
    "        input_nodes = []\n",
    "        output_nodes = []\n",
    "        connections = []\n",
    "\n",
    "        # create input and output\n",
    "        for i in range(num_inputs):\n",
    "            node = NodeGene(id=i, layer=0, activation='sigmoid', bias=random.uniform(-1, 1))\n",
    "            input_nodes.append(node)\n",
    "        for i in range(num_outputs):\n",
    "            index = i + num_inputs\n",
    "            node = NodeGene(id=index, layer=2, activation='sigmoid', bias=random.uniform(-1, 1))\n",
    "            output_nodes.append(node)\n",
    "        \n",
    "        # create connection between each input and output\n",
    "        for i in range(num_inputs):\n",
    "            for j in range(num_outputs):\n",
    "                in_node_id = i\n",
    "                out_node_id = j + num_inputs\n",
    "\n",
    "                innov_num = innov.get_inn_for_connection(in_node_id, out_node_id)\n",
    "                weight = random.uniform(-1, 1) # init random weight\n",
    "\n",
    "                connection = NodeConnection(id=innov_num, output_node_id=out_node_id, input_node_id=in_node_id, \n",
    "                        weight=weight, enabled=True, innovation=innov_num)\n",
    "                \n",
    "                connections.append(connection)\n",
    "\n",
    "        # create genome\n",
    "        all_nodes = input_nodes + output_nodes\n",
    "        genome = Genome(nodes=all_nodes, connections=connections)\n",
    "\n",
    "        return genome\n",
    "        \n",
    "    \"\"\"\n",
    "    create default population from genome state\n",
    "    \"\"\"\n",
    "    def init_population(self, pop_size, num_inputs, num_outputs, innov):\n",
    "        # initialize random population\n",
    "        population = []\n",
    "        for _ in range(pop_size):\n",
    "            genome = self.create_initial_genome(num_inputs, num_outputs, innov)\n",
    "            population.append(genome)\n",
    "        \n",
    "        return population\n",
    "    \n",
    "    def eval_population(self, population, max_steps, env):\n",
    "        for genome in population:\n",
    "            # run simulation episode\n",
    "            env.reset()\n",
    "\n",
    "            # perform actions at each timestamp\n",
    "            for step in range(max_steps):\n",
    "                obs = env.get_observation()\n",
    "                normalized_obs = self.normalize_observations(obs)\n",
    "                input_values = {i: normalized_obs[i+1] for i in range(5)}  # Maps 0->obs[1], 1->obs[2], etc.\n",
    "                action = self.forward_pass(genome, input_values)\n",
    "                obs, done = env.step(action)\n",
    "\n",
    "                max_indx = action.index(max(action))\n",
    "                print(f\"timestamp {step}: \"\n",
    "                    f\"Blue={obs[1]:.1f}, Red={obs[2]:.1f}, \"\n",
    "                    f\"Action= ({['Forward','Retreat','Attack','Hold'][max_indx]})\")\n",
    "                \n",
    "                if done:\n",
    "                    break\n",
    "                # viz.render()\n",
    "                # time.sleep(0.1)  # 100ms delay per step\n",
    "\n",
    "            # compute fitness of genome\n",
    "            final_obs = env.get_observation()\n",
    "            # normalized_obs = self.normalize_observations(final_obs)\n",
    "            genome.fitness = self.calculate_fitness(final_obs) # set fitness variable\n",
    "        \n",
    "    \n",
    "    def reproduce(self, population, innov):\n",
    "        new_population = []\n",
    "        pop_size = len(population)\n",
    "\n",
    "        # elitism: only keep top k performers based on fitness\n",
    "        # selection pressure: only allow top X% to reproduce\n",
    "        # these are the top performing strategies\n",
    "        FRAC = 0.8\n",
    "        NUM_ELITES = max(1, int(len(population) * FRAC))\n",
    "        elites = sorted(population, key=lambda g: g.fitness, reverse=True)[:NUM_ELITES]\n",
    "        new_population.extend(elites) \n",
    "\n",
    "        # TODO: only use population pool to preserve diversity\n",
    "        reproduction_pool = population\n",
    "\n",
    "        # reproduce from elites\n",
    "        while len(new_population) < pop_size:\n",
    "            parent1 = random.choice(reproduction_pool)\n",
    "            parent2 = random.choice(reproduction_pool)\n",
    "            \n",
    "            if parent1 != parent2:\n",
    "                print(\"parent1\", (parent1), \"\\nparent2\", (parent2))\n",
    "                child = crossover(parent1, parent2, reproduction_pool)\n",
    "                mutate_weights(child, mutation_rate=0.6, perturbation_strength=0.5, reset_rate=0.1)\n",
    "                mutate_connections(child, mutation_rate=0.6, innov=innov)\n",
    "                print(\"child\", child)\n",
    "                new_population.append(child)\n",
    "        \n",
    "        return new_population\n",
    "\n",
    "    def evolution_loop(self, pop_size, num_generations, env, num_inputs, num_outputs, innov):\n",
    "        # initialize & eval initial env\n",
    "        # we repeat the steps 100 times based on the timestamp\n",
    "        max_steps = 100 \n",
    "        population = self.init_population(pop_size, num_inputs, num_outputs, innov)\n",
    "        self.eval_population(population, max_steps, env)\n",
    "\n",
    "        # repeat for num of generations\n",
    "        for gen in range(num_generations):\n",
    "            self.eval_population(population, max_steps, env) # compute fitness of all genomes\n",
    "            \n",
    "            fitnesses = [g.fitness for g in population]\n",
    "            print(f\"Gen {gen+1}: max_fit={max(fitnesses):.4f}, \"\n",
    "                f\"avg_fit={sum(fitnesses)/len(fitnesses):.4f}, \"\n",
    "                f\"zero_fit (full fail)={sum(1 for f in fitnesses if f == 0)}\")\n",
    "\n",
    "            population = self.reproduce(population, innov) # assign back to population\n",
    "\n",
    "    # ==== HELPER ====\n",
    "    def normalize_observations(self, obs):\n",
    "        \"\"\"\n",
    "        normalize observations to [0, 1]\n",
    "        prevents saturating the sigmoid\n",
    "        \"\"\"\n",
    "        ranges = {\n",
    "            1: (0, 50),    # blue_units: expect 0-50\n",
    "            2: (0, 50),    # red_units: expect 0-50\n",
    "            3: (0, 1),     # wind_speed: already 0-1\n",
    "            4: (0, 1),     # blue_firepower: already 0-1\n",
    "            5: (0, 1),     # red_firepower: already 0-1\n",
    "        }\n",
    "\n",
    "        normalized = {}\n",
    "        for key, val in obs.items():\n",
    "            min, max = ranges[key]\n",
    "            normalized[key] = (val - min) / (max - min) if max > min else val\n",
    "        \n",
    "        return normalized\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13",
   "metadata": {},
   "source": [
    "## 6. test simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "from environment import WargameEnv\n",
    "from game import Game\n",
    "\n",
    "env = WargameEnv()\n",
    "viz = Game(env)\n",
    "\n",
    "# obs = env.get_observation()\n",
    "# action = loop.forward_pass(child, obs)\n",
    "# print(action)\n",
    "# env.step(action)\n",
    "# print(env.step(action))\n",
    "# viz.render()\n",
    "\n",
    "# init params\n",
    "innov = InnovationCounter(global_innovations=0, node_connection_innovations={})\n",
    "loop = EvolutionLoop(innov)\n",
    "\n",
    "num_inputs = 5\n",
    "num_outputs = 4 # up, down, left, wright\n",
    "pop_size = 60\n",
    "num_generations=50\n",
    "\n",
    "loop.evolution_loop(pop_size, num_generations, env, num_inputs, num_outputs, innov)\n",
    "viz.render()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15",
   "metadata": {},
   "source": [
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "myenv",
   "language": "python",
   "name": "myenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
